### ISSUE TITLE
Definition of compilation unit and crate is murky

### ISSUE THREAD

**Issue by mkaput** [starkware-libs/cairo#2890]
# Bug Report

**Cairo version: 1.0.0-alpha.7**

The documentation states that:

> A crate is a single compilation unit. It has a root directory, and a root module defined at the file lib.cairo under this directory.

But this is not true, because the compiler takes a set of crates as an input, so the compilation unit is actually a collection of crates.

It actually is pretty murky to me what `crate` is in context of Cairo.
- It basically is just a root module and is only considered from name resolution perspective; because Cairo compilation model means that basically all the code is compiled at once in a shared compiler database.
- The term `crate` is very Rust-specific and for this reason I went with calling it a `package` for Cairo in Scarb.

cc @cptartur @karol-bisztyga @MaksymilianDemitraszek @piotmag769

**Comment by spapinistarkware** [comment]
I think it's just a temporary limitation.
First note that also in rust, some code if other crates gets compiled (generic code).
Second, in the future we would link sierra code, IMO.

**Comment by spapinistarkware** [comment]
@mkaput anything to add? I think it's not very different from rust, and is inteded to be more similar in the future.
Can I close this?

**Comment by mkaput** [comment]
> I think it's just a temporary limitation.

I do not think this is just a temporary limitation, because its very existence has a big influence downstream. For example, adopting C compilation model (which Rust uses) would basically require rethinking many aspects of Scarb and package management model in general, because for example currently you cannot include two versions of same crate in a single compilation unit, and thus we live with single package ↔ single version ecosystem. Of course, I'm skipping implementation here because this can be changed, it's all about fundamentals, changing which would be a big breaking change.

Also, the compiler APIs do not express a possibility of linking at all, and as far as I see we are aiming towards stabilization of Cairo 1.0 soon, so changing that in the future would mean creating Cairo 2.0 in not long future?
